packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -65,199 +66,81 @@ class TimelineFlameChart extends StatelessWidget {
-  Widget build(BuildContext context) {
-    return LayoutBuilder(
-      builder: (context, constraints) {
-        return Scrollbar(
-          child: SingleChildScrollView(
-            controller: _scrollControllerX,
-            scrollDirection: Axis.horizontal,
-            child: Scrollbar(
-              child: SingleChildScrollView(
-                controller: _scrollControllerY,
-                scrollDirection: Axis.vertical,
-                child: _flameChartBody(constraints),
-              ),
-            ),
-          ),
-        );
-      },
-    );
-  }
-
-  Widget _flameChartBody(BoxConstraints constraints) {
-    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
-    final height = math.max(constraints.maxHeight, widget.height);
-
-    // TODO(kenz): rewrite this using slivers instead of a stack.
-    return ValueListenableBuilder(
-      valueListenable: widget.selectionNotifier,
-      builder: (context, selectedEvent, _) {
-        return Stack(
-          children: [
-            Container(
-              width: width,
-              height: height,
-            ),
-            ..._nodesInViewport(
-              constraints,
-              selectedEvent,
-            ), // pick what to show
-          ],
-        );
-      },
-    );
-  }
-
-  List<FlameChartNode> _nodesInViewport(
-    BoxConstraints constraints,
-    TimelineEvent selectedEvent,
-  ) {
-    // TODO(kenz): is creating all the FlameChartNode objects expensive even if
-    // we won't add them to the view? We create all the FlameChartNode objects
-    // and place them in FlameChart rows, but we only add [nodesInViewport] to
-    // the widget tree.
-    _buildFlameChartElements(selectedEvent);
-
-    // TODO(kenz): Use binary search method we use in html full timeline here.
-    final nodesInViewport = <FlameChartNode>[];
-    for (var row in rows) {
-      for (var node in row.nodes) {
-        final fitsHorizontally = node.rect.right >= scrollOffsetX &&
-            node.rect.left - scrollOffsetX <= constraints.maxWidth;
-        final fitsVertically = node.rect.bottom >= scrollOffsetY &&
-            node.rect.top - scrollOffsetY <= constraints.maxHeight;
-        if (fitsHorizontally && fitsVertically) {
-          nodesInViewport.add(node);
-        }
-      }
-    }
-    return nodesInViewport;
-  }
+  // Add one for the spacer offset between UI and GPU nodes.
+  int get gpuSectionStartRow =>
+      widget.data.uiEventFlow.depth +
+      _rowOffsetForTopTimeline +
+      _rowOffsetForSectionSpacer;
 
   // TODO(kenz): when optimizing this code, consider passing in the viewport
   // to only construct FlameChartNode elements that are in view.
-  void _buildFlameChartElements(TimelineEvent selectedEvent) {
+  @override
+  void initFlameChartElements() {
     _resetColorOffsets();
 
     rows = List.generate(
-      widget.data.uiEventFlow.depth + widget.data.gpuEventFlow.depth,
+      widget.data.uiEventFlow.depth +
```

Comment:
Suggestion: put this into a variable before using it.

Reply:
done